engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java

```diff
@@ -231,31 +240,185 @@ public ExtendedSearchIntegrator getExtendedSearchIntegrator() {
 	protected void validateSortFields(ExtendedSearchIntegrator extendedIntegrator, Iterable<Class<?>> targetedEntities) {
 		SortField[] sortFields = sort.getSort();
 		for ( SortField sortField : sortFields ) {
-			if ( sortField instanceof DistanceSortField ) {
-				validateDistanceSortField( extendedIntegrator, targetedEntities, sortField );
+			validateSortField( extendedIntegrator, targetedEntities, sortField );
+		}
+	}
+
+	private void validateSortField(ExtendedSearchIntegrator extendedIntegrator, Iterable<Class<?>> targetedEntities, SortField sortField) {
+		if ( sortField instanceof DistanceSortField ) {
+			validateDistanceSortField( extendedIntegrator, targetedEntities, sortField );
+		}
+		else if ( sortField.getType() != SortField.Type.CUSTOM ) {
+			if ( sortField.getField() == null ) {
+				validateNullSortField( sortField );
+			}
+			else {
+				validateCommonSortField( extendedIntegrator, targetedEntities, sortField );
+			}
+		}
+	}
+
+	private void validateNullSortField(SortField sortField) {
+		if ( sortField.getType() != SortField.Type.DOC && sortField.getType() != SortField.Type.SCORE ) {
+			throw LOG.sortRequiresIndexedField( sortField.getClass(), sortField.getField() );
+		}
+	}
+
+	private void validateDistanceSortField(ExtendedSearchIntegrator extendedIntegrator, Iterable<Class<?>> targetedEntities, SortField sortField) {
+		FieldDescriptor fieldDescriptor = findFieldDescriptor( extendedIntegrator, targetedEntities, sortField.getField() );
+		if ( fieldDescriptor == null ) {
+			throw LOG.sortRequiresIndexedField( sortField.getClass(), sortField.getField() );
+		}
+		if ( fieldDescriptor.getType() != Type.SPATIAL ) {
+			throw LOG.distanceSortRequiresSpatialField( sortField.getField() );
+		}
+	}
+
+	private FieldDescriptor findFieldDescriptor(ExtendedSearchIntegrator extendedIntegrator, Iterable<Class<?>> targetedEntities, String field) {
+		if ( field == null ) {
+			return null;
+		}
+		for ( Class<?> clazz : targetedEntities ) {
+			IndexedTypeDescriptor indexedTypeDescriptor = extendedIntegrator.getIndexedTypeDescriptor( clazz );
```

Comment:
Is there a reason why you used the FieldDescriptor API? From what I know, it's not recommended to use it in our code. See for instance Gunnar's comment in https://hibernate.atlassian.net/browse/HSEARCH-2070 .